pkg/server/home_workspaces.go

```diff
@@ -0,0 +1,661 @@
+		return found, retryLater, retryAfterSeconds, checkError
+	}
+
+	found = true
+	if !isHome {
+		return found, retryLater, retryAfterSeconds, checkError
+	}
+
+	if localWorkspace.Status.Phase != tenancyv1alpha1.ClusterWorkspacePhaseReady {
+		// We have to wait for the workspace to be Ready before allowing actions in it,
+		// but only for the the home workspaces.
+		// Waiting for home buckets to be ready is done in the `tryToCreate` function.
+
+		// TODO (david): For now we accept all requests in this workspace when it's initializing...
+		// TODO (david): In the future, we might have a way to identify requests that are done through the
+		// initializer virtual workspace
+		// Or it might be unnecessary since other requests will always be blocked until initialization is finished.
+		if localWorkspace.Status.Phase != tenancyv1alpha1.ClusterWorkspacePhaseInitializing {
+			retryLater = true
+			retryAfterSeconds = h.creationDelay
+			return found, retryLater, retryAfterSeconds, checkError
+		}
+	}
+
+	if userName == kuser.APIServerUser {
+		// if request user is system:apiserver, don't wait for the RBAC resources to be setup
+		// (the system:apiserver user request is probably being creating those RBAC resources)
+		return found, retryLater, retryAfterSeconds, checkError
+	}
+
+	// For home workspaces, also wait for related RBAC resources to be setup,
+	// unless it's a system:apiserver user request
+	if rbacFound, searchRBACError := h.searchHomeWorkspaceRBACLocally(workspaceName, userName); searchRBACError != nil {
+		found = false
+		retryLater = false
+		retryAfterSeconds = 0
+		checkError = searchRBACError
+	} else if !rbacFound {
+		// Retry sooner than the creation delay, because it's probably a question of
+		// the local informer cache not being up-to-date, or the request having been sent
+		// to the wrong shard.
+		// Retrying quicly should be sufficient to fix it.
+		retryAfterSeconds = 1
+		retryLater = true
+	}
+	return found, retryLater, retryAfterSeconds, checkError
+}
+
+// tryToCreate tries to create the home or home bucket workspace, with an external client,
+// if it has not been found in the local informers.
+// Of course it can be that it has been created in the meantime (either concurrently on the current shard or
+// on another shard). We don't error in this case.
+// If it cannpot be created because the parent home bucket doesn't exist, create the parent first and
+// retry the creation of the current workspace later.
+// When creating a Home workspace, also create the related RBAC resources.
+func (h *homeWorkspaceHandler) tryToCreate(ctx context.Context, userName string, workspaceToCheck logicalcluster.Name, workspaceType tenancyv1alpha1.ClusterWorkspaceTypeName) (retryLater bool, retryAfterSeconds int, createError error) {
```

Comment:
s/createError/err/

And please (nearly) always do explicit `return ..., err`. Relying on return variable name is good for documentation, but fragile.

Reply:
Left this return variable name, but I don't use it in the code anymore.